server/src/main/java/io/spine/examples/shareaware/server/investment/InvestmentAggregate.java

```diff
@@ -57,4 +64,68 @@ private void event(SharesAdded e) {
                 .setId(e.getInvestment())
                 .setSharesAvailable(newAvailableShares);
     }
+
+    @Assign
+    SharesReserved on(ReserveShares c) throws InsufficientShares {
+        if (state().getSharesAvailable() < c.getQuantity()) {
+            throw InsufficientShares
+                    .newBuilder()
+                    .setInvestment(c.getInvestment())
+                    .setProcess(c.getProcess())
+                    .setQuantity(c.getQuantity())
+                    .build();
+        }
+        return SharesReserved
+                .newBuilder()
+                .setInvestment(c.getInvestment())
+                .setProcess(c.getProcess())
+                .setQuantity(c.getQuantity())
+                .vBuild();
+    }
+
+    @Apply
+    private void event(SharesReserved e) {
+        int newAvailableShares = state().getSharesAvailable() - e.getQuantity();
+        String saleId = e.getProcess()
+                         .getUuid();
+        builder()
+                .setSharesAvailable(newAvailableShares)
+                .putSharesReserved(saleId, e.getQuantity());
+    }
+
+    @Assign
+    SharesReservationCompleted on(CompleteSharesReservation c) {
+        return SharesReservationCompleted
+                .newBuilder()
+                .setInvestment(c.getInvestment())
+                .setProcess(c.getProcess())
+                .setSharesAvailable(state().getSharesAvailable())
+                .vBuild();
+    }
+
+    @Apply
+    private void event(SharesReservationCompleted e) {
+        builder().removeSharesReserved(e.getProcess()
```

Comment:
When emitted, `SharesReservationCompleted` includes `shares_available` field. I am not sure why we don't use it in the applier. Or, if it is not needed in this applier, why we have this field in this event at all.

Something does not quite add up here to me.

Reply:
But the information about available shares is needed for the `HeldShares` projection and I cannot get this information from anywhere else. Of course, I can calculate the amount of the currently available shares in the projection, but you told me not to do it in some previous PR.